addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/discovery/RioSystemDiscovery.java

```diff
@@ -0,0 +1,188 @@
+import org.openhab.binding.russound.internal.net.SocketChannelSession;
+import org.openhab.binding.russound.internal.net.SocketSession;
+import org.openhab.binding.russound.internal.net.WaitingSessionListener;
+import org.openhab.binding.russound.internal.rio.RioConstants;
+import org.openhab.binding.russound.internal.rio.system.RioSystemConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * This implementation of {@link DiscoveryService} will scan the network for any Russound RIO system devices. The scan
+ * will occur against all network interfaces.
+ *
+ * @author Tim Roberts
+ *
+ */
+public class RioSystemDiscovery extends AbstractDiscoveryService {
+    /** The logger */
+    private final static Logger logger = LoggerFactory.getLogger(RioSystemDiscovery.class);
+
+    /**
+     * Creates the system discovery service looking for {@link RioConstants#BRIDGE_TYPE_RIO}. The scan will take at most
+     * 120 seconds (depending on how many network interfaces there are)
+     */
+    public RioSystemDiscovery() {
+        super(ImmutableSet.of(RioConstants.BRIDGE_TYPE_RIO), 120);
+    }
+
+    /**
+     * Starts the scan. For each network interface (that is up and not a loopback), all 255 addresses will be iterated
+     * and checked for something open on port 9621. If that port is open, a russound controller "type" command will be
+     * issued. If the response is a correct pattern, we assume it's a rio system device and will emit a
+     * {{@link #thingDiscovered(DiscoveryResult)}
+     */
+    @Override
+    protected void startScan() {
+        Enumeration<NetworkInterface> interfaces;
+        try {
+            interfaces = NetworkInterface.getNetworkInterfaces();
+        } catch (SocketException e1) {
+            return;
+        }
+
+        while (interfaces.hasMoreElements()) {
+            final NetworkInterface networkInterface = interfaces.nextElement();
+            try {
+                if (networkInterface.isLoopback() || !networkInterface.isUp()) {
+                    continue;
+                }
+            } catch (SocketException e) {
+                continue;
+            }
+
+            final Enumeration<InetAddress> inetAddresses = networkInterface.getInetAddresses();
+            while (inetAddresses.hasMoreElements()) {
+                final InetAddress inetAddress = inetAddresses.nextElement();
+
+                if (inetAddress instanceof Inet6Address) {
+                    continue;
+                }
+
+                final String hostAddress = inetAddress.getHostAddress();
+                final int lastDot = hostAddress.lastIndexOf('.');
+                final String subnet = hostAddress.substring(0, lastDot + 1);
+                logger.trace("Starting scan of {}", subnet);
+
+                final String[] addresses = new String[254];
+                for (int i = 1; i < 255; i++) {
```

Comment:
Are you aware that subnets can also be bigger?

You can have this complete range in a single home network: 192.168.0.0 - 192.168.255.255 or even bigger.

Besides that maybe a warning should be added before starting discovery like this, because some (company) networks might not like this port-scan. And if someone is connected to a vpn that network will be scanned as well.